compiler/rustc_typeck/src/check/gather_locals.rs

```diff
@@ -6,15 +6,24 @@ use rustc_infer::infer::type_variable::{TypeVariableOrigin, TypeVariableOriginKi
 use rustc_middle::ty::Ty;
 use rustc_span::Span;
 use rustc_trait_selection::traits;
+use std::mem;
 
 pub(super) struct GatherLocalsVisitor<'a, 'tcx> {
     fcx: &'a FnCtxt<'a, 'tcx>,
     parent_id: hir::HirId,
+    // parameters are special cases of patterns, but we want to handle them as
+    // *distinct* cases. so track when we are hitting a pattern *within* an fn
+    // parameter.
+    outermost_fn_param_pat: bool,
 }
 
 impl<'a, 'tcx> GatherLocalsVisitor<'a, 'tcx> {
-    pub(super) fn new(fcx: &'a FnCtxt<'a, 'tcx>, parent_id: hir::HirId) -> Self {
-        Self { fcx, parent_id }
+    pub(super) fn new(
+        fcx: &'a FnCtxt<'a, 'tcx>,
+        parent_id: hir::HirId,
+        outermost_fn_param_pat: bool,
+    ) -> Self {
+        Self { fcx, parent_id, outermost_fn_param_pat }
```

Comment:
Wouldn't it make more sense to always set `outermost_fn_param_pat: false`? Seems like some internal tracking that we wouldn't want the outside world to control.